Introduced Global checkpoints for Sequence Numbers

bleskes opened the pull request:
This is a work in progress PR for introducing Global Checkpoints into the feature/seq_no branch.  I made it to get some feedback before things are finalized / cycles are spent on docs etc. It is fairly solid, but a couple of things miss:

1) Lucene level doc rejections are not yet dealt with (I need to do some work on master).
2) Nothing is persisted yet. This will be a future PR.
3) No tests - I'll add those if we're good on the setup.

bleskes commented:
@jasontedor  I fixed the issue we ran into yesterday where the a primary relocation target started updating the global checkpoint too early

jasontedor commented on core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java:
```diff
@@ -864,10 +883,15 @@ void performOnReplica(final ShardRouting shard, final String nodeId) {
-                        @Override
-                        public void handleResponse(TransportResponse.Empty vResponse) {
-                            onReplicaSuccess();
+            transportService.sendRequest(node, transportReplicaAction, replicaRequest, transportOptions, new TransportResponseHandler<ShardResponse>() {
```
Nit: the indentation is off throughout the body of the anonymous `TransportRequestHandler`.

jasontedor commented on core/src/main/java/org/elasticsearch/cluster/metadata/MetaDataCreateIndexService.java:
```diff
@@ -299,7 +305,7 @@ public ClusterState execute(ClusterState currentState) throws Exception {
                     final IndexMetaData tmpImd = IndexMetaData.builder(request.index()).settings(actualIndexSettings).build();
                     // create the index here (on the master) to validate it can be created, as well as adding the mapping
-                    indicesService.createIndex(nodeServicesProvider, tmpImd, Collections.emptyList());
+                    indicesService.createIndex(nodeServicesProvider, tmpImd, Collections.emptyList(),shardId -> {});
```
Nit: add a space after the `,` before `shardId`: `,shardId` -> `, shardId`

jasontedor commented on core/src/main/java/org/elasticsearch/cluster/metadata/MetaDataIndexAliasesService.java:
```diff
@@ -99,7 +103,8 @@ public ClusterState execute(final ClusterState currentState) {
                                         try {
-                                            indexService = indicesService.createIndex(nodeServicesProvider, indexMetaData, Collections.emptyList());
+                                            indexService = indicesService.createIndex(nodeServicesProvider, indexMetaData, Collections.emptyList(), shardId -> {
+                                            });
```
Nit: wrap this to the preceding line.

jasontedor commented on core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java:
```diff
@@ -989,35 +1023,74 @@ protected boolean shouldExecuteReplication(Settings settings) {
     }
 
+    protected static class ShardResponse extends ActionResponse {
+        long localCheckpoint;
```
I think that this can have private accessibility.

bleskes replied:
opted with making the class private and leave these as is they are accessed from the outside. I know it's not needed, but I feel it's clearer.

jasontedor commented on core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java:
```diff
@@ -989,35 +1023,74 @@ protected boolean shouldExecuteReplication(Settings settings) {
 
+    protected static class ShardResponse extends ActionResponse {
+        long localCheckpoint;
+        public String allocationId;
```
I think that only private accessibility is needed here?

jasontedor commented on core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java:
```diff
@@ -989,35 +1023,74 @@ protected boolean shouldExecuteReplication(Settings settings) {
         private final AtomicBoolean closed = new AtomicBoolean();
         private final long opPrimaryTerm;
 
         IndexShardReferenceImpl(IndexShard counter) {
```
Change the name of the constructor parameter to `indexShard` too?

bleskes replied:
done

jasontedor commented on core/src/main/java/org/elasticsearch/cluster/metadata/MetaDataMappingService.java:
```diff
@@ -129,7 +140,8 @@ ClusterState executeRefresh(final ClusterState currentState, final List<RefreshT
                 // we need to create the index here, and add the current mapping to it, so we can merge
-                indexService = indicesService.createIndex(nodeServicesProvider, indexMetaData, Collections.emptyList());
+                indexService = indicesService.createIndex(nodeServicesProvider, indexMetaData, Collections.emptyList(), shardId -> {
+                });
```
Nit: wrap this to the preceding line.

bleskes replied:
I set it back, but Intellij likes it that way and I couldn't find a way to convince it otherwise without making other stuff ugly.

jasontedor replied:
Well, that's annoying but I can understand not wanting to fight with the tooling.

jasontedor commented on core/src/main/java/org/elasticsearch/cluster/metadata/MetaDataMappingService.java:
```diff
@@ -210,7 +222,8 @@ public void refreshMapping(final String index, final String indexUUID) {
                             indicesToClose.add(index);
-                            IndexService indexService = indicesService.createIndex(nodeServicesProvider, indexMetaData, Collections.emptyList());
+                            IndexService indexService = indicesService.createIndex(nodeServicesProvider, indexMetaData, Collections.emptyList(), shardId -> {
+                            });
```
Nit: wrap this to the preceding line.

jasontedor commented on core/src/main/java/org/elasticsearch/cluster/routing/IndexShardRoutingTable.java:
```diff
@@ -214,6 +221,16 @@ public int getSize() {
+        return this.allInitializingShards;
+    }
+
+
```
Nit: there is an extra blank line here. Sorry.

jasontedor commented on core/src/main/java/org/elasticsearch/index/seqno/SeqNoStats.java:
```diff
@@ -51,15 +53,20 @@ public long getLocalCheckpoint() {
         return localCheckpoint;
     }
 
+    public long getGlobalCheckpoint() {
```
I assume that this will be used in yet-to-be-written tests?

bleskes replied:
yeah, I will add some integration tests.

jasontedor commented on core/src/main/java/org/elasticsearch/index/seqno/SeqNoStats.java:
```diff
@@ -51,15 +53,20 @@ public long getLocalCheckpoint() {
     @Override
     public SeqNoStats readFrom(StreamInput in) throws IOException {
-        return new SeqNoStats(in.readLong(), in.readLong());
+        return new SeqNoStats(in.readZLong(), in.readZLong(), in.readZLong());
```
It looks like we both missed this on the local checkpoint pull request?

jasontedor commented on core/src/main/java/org/elasticsearch/index/shard/ShadowIndexShard.java:
```diff
@@ -46,7 +46,8 @@
                             IndexEventListener indexEventListener, IndexSearcherWrapper wrapper, NodeServicesProvider provider) throws IOException {
-        super(shardId, indexSettings, path, store, indexCache, mapperService, similarityService, indexFieldDataService, engineFactory, indexEventListener, wrapper, provider);
+        super(shardId, indexSettings, path, store, indexCache, mapperService, similarityService, indexFieldDataService, engineFactory, indexEventListener, wrapper, provider, (shard) -> {
+        });
```
Nit: wrap this to the previous line.

jasontedor commented on core/src/test/java/org/elasticsearch/index/IndexModuleTests.java:
```diff
@@ -315,7 +327,8 @@ public void testDefaultQueryCacheImplIsSelected() throws IOException {
         IndexModule module = new IndexModule(IndexSettingsModule.newIndexSettings(new Index("foo"), indexSettings), null, new AnalysisRegistry(null, environment));
-        IndexService indexService = module.newIndexService(nodeEnvironment, deleter, nodeServicesProvider, mapperRegistry);
+        IndexService indexService = module.newIndexService(nodeEnvironment, deleter, nodeServicesProvider, mapperRegistry, shardId -> {
+        });
```
Nit: wrap to the preceding line (and on lines 141, 155, 180, 211, 239, 257, 272 and 320).

bleskes replied:
all auto formatted. See comment from above

jasontedor commented on core/src/test/java/org/elasticsearch/indices/IndicesLifecycleListenerSingleNodeTests.java:
```diff
@@ -92,7 +92,8 @@ public void afterIndexShardDeleted(ShardId shardId, Settings indexSettings) {
             NodeServicesProvider nodeServicesProvider = getInstanceFromNode(NodeServicesProvider.class);
-            IndexService index = indicesService.createIndex(nodeServicesProvider, metaData, Arrays.asList(countingListener));
+            IndexService index = indicesService.createIndex(nodeServicesProvider, metaData, Arrays.asList(countingListener), shardId -> {
+            });
```
Nit: wrap to the preceding line.

jasontedor commented on core/src/main/java/org/elasticsearch/index/seqno/GlobalCheckpointService.java:
```diff
@@ -0,0 +1,199 @@
+
+    /**
+     * This map holds the last known local checkpoint for every initializing shard copy that is still undergoing recovery.
+     * These shards <strong>do not</strong> participate in determining the global checkpoint. These map is needed to make sure that when
```
Typo: `These map` -> `This map`.

jasontedor commented on core/src/main/java/org/elasticsearch/index/seqno/GlobalCheckpointService.java:
```diff
@@ -0,0 +1,199 @@
+    }
+
+    private boolean updateLocalCheckPointInMap(String allocationId, long localCheckpoint, ObjectLongMap<String> checkpointsMap, String name) {
+        int indexOfKey = checkpointsMap.indexOf(allocationId);
```
Assert that the current thread holds a lock on `this`? The results from `ObjectLongMap#indexOf` remain valid only if no one else is mutating.

jasontedor commented on core/src/main/java/org/elasticsearch/index/seqno/GlobalCheckpointService.java:
```diff
@@ -0,0 +1,199 @@
+        logger.trace("added local checkpoint of [{}] ([{}]) to the tracking map", allocationId, localCheckpoint);
+    }
+
+    private boolean updateLocalCheckPointInMap(String allocationId, long localCheckpoint, ObjectLongMap<String> checkpointsMap, String name) {
```
Nit: Change the casing of `CheckPoint` to `Checkpoint`.

jasontedor commented on core/src/main/java/org/elasticsearch/index/seqno/GlobalCheckpointService.java:
```diff
@@ -0,0 +1,199 @@
+    /**
+     * Scans through the currently known local checkpoint and updates the global check point accordingly.
+     *
+     * @return true if the checkpoint has been updated or if it can not be update since one of the local checkpoint
```
Typo: "`local checkpoint`" -> "`local checkpoints`".

jasontedor replied:
Typo: "`can not be update`" -> "`can not be updated`".

jasontedor commented on core/src/main/java/org/elasticsearch/index/seqno/GlobalCheckpointService.java:
```diff
@@ -0,0 +1,199 @@
+    }
+
+    /**
+     * Scans through the currently known local checkpoint and updates the global check point accordingly.
```
Typo: "`check point`" -> "`checkpoint`".

jasontedor commented on core/src/main/java/org/elasticsearch/index/seqno/SequenceNumbersService.java:
```diff
@@ -54,6 +59,59 @@ public void markSeqNoAsCompleted(long seqNo) {
+    }
+
+    /**
+     * Scans through the currently known local checkpoint and updates the global check point accordingly.
```
Typo: "`check point`" -> "`checkpoint`".

jasontedor commented on core/src/main/java/org/elasticsearch/index/seqno/SequenceNumbersService.java:
```diff
@@ -54,6 +59,59 @@ public void markSeqNoAsCompleted(long seqNo) {
+    /**
+     * Scans through the currently known local checkpoint and updates the global check point accordingly.
+     *
+     * @return true if the checkpoint has been updated or if it can not be update since one of the local checkpoint
```
Typo: "`local checkpoint`" -> "`local checkpoints`".

jasontedor replied:
Typo: "`can not be update`" -> "`can not be updated`".

jasontedor commented on core/src/main/java/org/elasticsearch/index/seqno/GlobalCheckpointSyncAction.java:
```diff
@@ -0,0 +1,117 @@
+        execute(new ReplicationRequest(shardId), new ActionListener<ReplicationResponse>() {
+            @Override
+            public void onResponse(ReplicationResponse replicationResponse) {
+                logger.trace("{} global check point successfully updated (shard info [{}])", shardId, replicationResponse.getShardInfo());
```
Since this is in the common code path and because of `ShardInfo#toString` using `XContent` to generate its value, I think this should be wrapped in a `logger#isTraceEnabled` check.

bleskes replied:
This is called every 30s, I don't think we should worry too much about it, but I'll change.

jasontedor commented on core/src/main/java/org/elasticsearch/index/seqno/GlobalCheckpointSyncAction.java:
```diff
@@ -0,0 +1,117 @@
+
+    final static class ReplicaRequest extends ReplicationRequest<GlobalCheckpointSyncAction.ReplicaRequest> {
+
+        long checkPoint;
```
Nit: can you rename this to `checkpoint` for consistency?

jasontedor replied:
Can this field be private?

bleskes replied:
I actually made it public. Same reasoning as above. Sadly the class can't be made private.

jasontedor commented on core/src/main/java/org/elasticsearch/index/seqno/GlobalCheckpointSyncAction.java:
```diff
@@ -0,0 +1,117 @@
+        private ReplicaRequest() {
+        }
+
+        public ReplicaRequest(ReplicationRequest primaryRequest, long checkPoint) {
```
Nit: can you rename the constructor parameter to `checkpoint` for consistency?

jasontedor commented on core/src/main/java/org/elasticsearch/index/seqno/GlobalCheckpointSyncAction.java:
```diff
@@ -0,0 +1,117 @@
+    protected Tuple<ReplicationResponse, ReplicaRequest> shardOperationOnPrimary(MetaData metaData, ReplicationRequest request) throws Throwable {
+        IndexService indexService = indicesService.indexServiceSafe(request.shardId().getIndex());
+        IndexShard indexShard = indexService.getShard(request.shardId().id());
+        long checkPoint = indexShard.getGlobalCheckpoint();
```
Nit: can you rename this to `checkpoint` for consistency?

jasontedor commented on core/src/main/java/org/elasticsearch/index/seqno/SequenceNumbersService.java:
```diff
@@ -54,6 +59,59 @@ public void markSeqNoAsCompleted(long seqNo) {
+    /**
+     * updates the global checkpoint on a replica shard (after it has been updated by the primary).
+     */
+    public void updateGlobalCheckpointOnReplica(long checkPoint) {
```
Nit: `checkpoint`. :)

jasontedor commented on core/src/main/java/org/elasticsearch/index/shard/IndexShard.java:
```diff
@@ -1297,11 +1332,188 @@ public TimeValue getInactiveTime() {
+    /**
+     * updates the global checkpoint on a replica shard (after it has been updated by the primary).
+     */
+    public void updateGlobalCheckpointOnReplica(long checkPoint) {
```
Nit: `checkpoint`. :)

jasontedor commented on core/src/main/java/org/elasticsearch/index/seqno/GlobalCheckpointService.java:
```diff
@@ -0,0 +1,199 @@
+     * All shard copies in this map participate in determining the global checkpoint
+     * keyed by allocation ids
+     */
+    final ObjectLongMap<String> activeLocalCheckpoints;
```
I think that this can be private?

jasontedor commented on core/src/main/java/org/elasticsearch/index/seqno/GlobalCheckpointService.java:
```diff
@@ -0,0 +1,199 @@
+     * to speed through recovery. These shards are treated as valid copies and participate in determining the global checkpoint.
+     * Keyed by allocation ids
+     */
+    final ObjectLongMap<String> inSyncLocalCheckpoints; // keyed by allocation ids
```
I think that this can be private?

jasontedor commented on core/src/main/java/org/elasticsearch/index/seqno/GlobalCheckpointService.java:
```diff
@@ -0,0 +1,199 @@
+     * while recovering the shard.
+     * Keyed by allocation ids
+     */
+    final ObjectLongMap<String> trackingLocalCheckpoint;  // keyed by allocation ids
```
I think that this can be private?

jasontedor commented on core/src/main/java/org/elasticsearch/index/seqno/GlobalCheckpointService.java:
```diff
@@ -0,0 +1,199 @@
+     */
+    final ObjectLongMap<String> trackingLocalCheckpoint;  // keyed by allocation ids
+
+    long globalCheckpoint = SequenceNumbersService.UNASSIGNED_SEQ_NO;
```
I think that this can be private?

jasontedor commented on core/src/test/java/org/elasticsearch/index/IndexModuleTests.java:
```diff
@@ -133,7 +137,8 @@ public void testWrapperIsBound() throws IOException {
         module.engineFactory.set(new MockEngineFactory(AssertingDirectoryReader.class));
-        IndexService indexService = module.newIndexService(nodeEnvironment, deleter, nodeServicesProvider, mapperRegistry);
+        IndexService indexService = module.newIndexService(nodeEnvironment, deleter, nodeServicesProvider, mapperRegistry, shardId -> {
+        });
```
Nit: wrap this to the preceding line.

jasontedor commented on core/src/test/java/org/elasticsearch/index/IndexModuleTests.java:
```diff
@@ -146,7 +151,8 @@ public void testRegisterIndexStore() throws IOException {
         module.addIndexStore("foo_store", FooStore::new);
-        IndexService indexService = module.newIndexService(nodeEnvironment, deleter, nodeServicesProvider, mapperRegistry);
+        IndexService indexService = module.newIndexService(nodeEnvironment, deleter, nodeServicesProvider, mapperRegistry, shardId -> {
+        });
```
Nit: wrap this to the preceding line.

jasontedor commented on core/src/test/java/org/elasticsearch/index/IndexModuleTests.java:
```diff
@@ -170,7 +176,8 @@ public void beforeIndexDeleted(IndexService indexService) {
         module.addIndexEventListener(eventListener);
-        IndexService indexService = module.newIndexService(nodeEnvironment, deleter, nodeServicesProvider, mapperRegistry);
+        IndexService indexService = module.newIndexService(nodeEnvironment, deleter, nodeServicesProvider, mapperRegistry, shardId -> {
+        });
```
Nit: wrap this to the preceding line.

jasontedor commented on core/src/test/java/org/elasticsearch/index/IndexModuleTests.java:
```diff
@@ -227,7 +235,8 @@ public Similarity get() {
 
-        IndexService indexService = module.newIndexService(nodeEnvironment, deleter, nodeServicesProvider, mapperRegistry);
+        IndexService indexService = module.newIndexService(nodeEnvironment, deleter, nodeServicesProvider, mapperRegistry, shardId -> {
+        });
```
Nit: wrap this to the preceding line.

jasontedor commented on core/src/test/java/org/elasticsearch/index/IndexModuleTests.java:
```diff
@@ -244,7 +253,8 @@ public void testSetupUnknownSimilarity() throws IOException {
         try {
-            module.newIndexService(nodeEnvironment, deleter, nodeServicesProvider, mapperRegistry);
+            module.newIndexService(nodeEnvironment, deleter, nodeServicesProvider, mapperRegistry, shardId -> {
+            });
```
Nit: wrap this to the preceding line.

jasontedor commented on core/src/test/java/org/elasticsearch/index/IndexModuleTests.java:
```diff
@@ -258,7 +268,8 @@ public void testSetupWithoutType() throws IOException {
         try {
-            module.newIndexService(nodeEnvironment, deleter, nodeServicesProvider, mapperRegistry);
+            module.newIndexService(nodeEnvironment, deleter, nodeServicesProvider, mapperRegistry, shardId -> {
+            });
```
Nit: wrap this to the preceding line.

jasontedor commented on core/src/test/java/org/elasticsearch/index/IndexModuleTests.java:
```diff
@@ -305,7 +316,8 @@ public void testRegisterCustomQueryCache() throws IOException {
 
-        IndexService indexService = module.newIndexService(nodeEnvironment, deleter, nodeServicesProvider, mapperRegistry);
+        IndexService indexService = module.newIndexService(nodeEnvironment, deleter, nodeServicesProvider, mapperRegistry, shardId -> {
+        });
```
Nit: wrap this to the preceding line.

jasontedor commented on core/src/main/java/org/elasticsearch/index/shard/IndexShard.java:
```diff
@@ -1297,11 +1332,188 @@ public TimeValue getInactiveTime() {
+        }
+
+        /** returns true if this task should be running for this shard and setting */
+        protected boolean shouldRun() {
```
I think that I would prefer for this to be private, and chain a protected `internalShouldRun` that can be overridden on the end of the conjunction:

```java
return interval.millis() > 0 && shouldRunStates().contains(state) && shardRouting.primary() && internalShouldRun();
```

bleskes replied:
to clarify are you trying to have the mutex lock acquired once ? the way it's done the control is in the child and I think that's correct?

jasontedor replied:
Nope. Instead what I'm looking at is the logic in the base `shouldRun` method and I think that as it things are designed currently, that logic should always be used and so I just see it simpler to not allow derived classes to override.

bleskes replied:
I think it's simpler now and gives more control to the sub class. Unless you feel strongly about it I prefer to keep as is.

jasontedor replied:
Asking out of curiosity to understand your thinking: why does the subclass need such control here? By the way, I do not feel strongly. :)

jasontedor commented on core/src/main/java/org/elasticsearch/index/shard/IndexShard.java:
```diff
@@ -1297,11 +1332,188 @@ public TimeValue getInactiveTime() {
+            }
+        }
+
+        final EnumSet<IndexShardState> DEFAULT_RUNNING_STATES = EnumSet.of(IndexShardState.RECOVERING,
```
Can you double check that `IndexShardState.CREATED` is not needed here too; the `EngineRefresher` use to be eligible to run in that state too?

jasontedor commented:
Conceptually this is very sound, I only have a bunch of nits and minor comments and otherwise LGTM.

jasontedor commented on core/src/main/java/org/elasticsearch/index/shard/IndexShard.java:
```diff
@@ -1297,11 +1332,188 @@ public TimeValue getInactiveTime() {
+        @Override
+        protected boolean shouldRun() {
+            synchronized (mutex) {
+                return super.shouldRun() && shardRouting.primary();
```
I don't understand this condition. The base implementation already includes ``&& shardRouting.primary()``?

bleskes replied:
the bug is in the base class. IT should not be tied to a primary. Good catch.

bleskes commented:
@jasontedor I rebased this and added a very basic integration test. It's hard to write heavier IT tests with all the rest of the system being wired up (recoveries, engine persistency etc.). I want to got through it myself again - it's been a long time since this was written - but I think it's ready to start the review process again.

bleskes commented:
@jasontedor I went through the code again and cleaned up some minor things. It's ready for review, when you have time.

jasontedor commented on core/src/main/java/org/elasticsearch/index/seqno/GlobalCheckpointService.java:
```diff
@@ -0,0 +1,245 @@
+
+/**
+ * A shard component that is responsible of tracking the global checkpoint. The global checkpoint
+ * is the highest seq_no for which all lower (or equal) seq no have been process on all shards that
```
Nit: "seq no" -> "seq_no" (inconsistencies will make searching more difficult)

jasontedor replied:
Nit: "process" -> "processed"

bleskes replied:
done

jasontedor commented on core/src/main/java/org/elasticsearch/index/seqno/GlobalCheckpointService.java:
```diff
@@ -0,0 +1,245 @@
+     * while recovering the shard.
+     * Keyed by allocation ids
+     */
+    final private ObjectLongMap<String> trackingLocalCheckpoint;  // keyed by allocation ids
```
Duplicate documentation of being keyed by allocation ids, here and in the above Javadoc.

bleskes replied:
left over indeed. removed.

jasontedor commented on core/src/main/java/org/elasticsearch/index/IndexService.java:
```diff
@@ -118,6 +120,7 @@
     private volatile AsyncTranslogFSync fsyncTask;
     private final ThreadPool threadPool;
     private final BigArrays bigArrays;
+    private volatile AsyncGlobalCheckpointTask globalCheckpointTask;
```
It seems like this task is never executed?

jasontedor replied:
I see, it's scheduled from the constructor of `BaseAsyncTask`. I do not think `volatile` is needed here then, and this field can be marked `final`.

bleskes replied:
correct! good catch.

jasontedor commented on core/src/main/java/org/elasticsearch/index/seqno/GlobalCheckpointService.java:
```diff
@@ -0,0 +1,245 @@
+     */
+    synchronized public boolean updateCheckpointOnPrimary() {
+        long minCheckpoint = Long.MAX_VALUE;
+        if (activeLocalCheckpoints.isEmpty() &&  inSyncLocalCheckpoints.isEmpty()) {
```
Nit: there is an extra space after the `&&` and before `inSyncLocalCheckpoints`.

bleskes replied:
removed

jasontedor commented on core/src/main/java/org/elasticsearch/index/IndexService.java:
```diff
@@ -718,6 +725,33 @@ private void maybeRefreshEngine() {
     }
 
+    private void maybeUpdateGlobalCheckpoints() {
+        if (indexSettings.getGlobalCheckpointInterval().millis() > 0) {
```
Is this guard really needed? The task only gets scheduled if the interval is positive. The only situation this guards against is the setting being dynamically changed to zero or -1from a positive value and a task was already scheduled?

bleskes replied:
yeah, this is copy paste from the refresh logic which has to be defensive for dynamic setting update. Not needed here indeed (yet :)) . I removed it.

jasontedor commented on core/src/main/java/org/elasticsearch/index/IndexSettings.java:
```diff
@@ -106,6 +106,9 @@
 
+    public static final Setting<TimeValue> INDEX_SEQ_NO_CHECKPOINT_SYNC_INTERVAL =
+        Setting.timeSetting("index.seq_no.checkpoint_sync_interval", new TimeValue(30, TimeUnit.SECONDS),
+            new TimeValue(-1, TimeUnit.MILLISECONDS), Property.Dynamic, Property.IndexScope);
```
What is the significance of -1? Why is the minimum not just zero?

bleskes replied:
It allows you set it to a negative value to disable the updates. Not sure we need it yet (probably will need to have it for testing at some point.

bleskes replied:
It allows you set it to a negative value to disable the updates. Not sure we need it yet (probably will need to have it for testing at some point.

jasontedor replied:
Doesn't zero achieve the same thing?

bleskes replied:
well, I think -1 is more intuitive than 0. In general we can maybe invest in setting parsing that accept -1 for disabling and requires a strictly positive value o.w. . This is however a different change IMO - here I just copied existing patterns from, for example, the refresh interval.

jasontedor replied:
Okay, I see what you mean. As interval tends to zero, the frequency of updates increases until zero when there are none. You're right it would be better in this case to disallow zero.

jasontedor replied:
I also agree that doesn't need to be addressed in this PR.

jasontedor commented on core/src/main/java/org/elasticsearch/indices/cluster/IndicesClusterStateService.java:
```diff
@@ -379,7 +384,9 @@ private void applyNewIndices(final ClusterChangedEvent event) {
-                    indicesService.createIndex(nodeServicesProvider, indexMetaData, buildInIndexListener);
+                    indicesService.createIndex(nodeServicesProvider, indexMetaData, buildInIndexListener,
+                        globalCheckpointSyncAction::updateCheckpointForShard
+                        );
```
Nit: `);` can be on the previous line

bleskes replied:
fixed

jasontedor commented on core/src/main/java/org/elasticsearch/index/seqno/GlobalCheckpointService.java:
```diff
@@ -0,0 +1,245 @@
+     */
+    synchronized public void markAllocationIdAsInSync(String allocationId, long localCheckpoint) {
+        if (trackingLocalCheckpoint.containsKey(allocationId) == false) {
+            // master have change it's mind and removed this allocation, ignore.
```
Nit: "have change it's" -> "has changed its"

bleskes replied:
fixed

jasontedor commented on core/src/main/java/org/elasticsearch/index/seqno/GlobalCheckpointService.java:
```diff
@@ -0,0 +1,245 @@
+        for (String activeId : activeAllocationIds) {
+            if (activeLocalCheckpoints.containsKey(activeId) == false) {
+                long knownCheckpoint = trackingLocalCheckpoint.getOrDefault(activeId,
+                    SequenceNumbersService.UNASSIGNED_SEQ_NO);
```
Nit: this does not need to be on a separate line

bleskes replied:
fixed.

jasontedor commented on core/src/main/java/org/elasticsearch/index/seqno/GlobalCheckpointService.java:
```diff
@@ -0,0 +1,245 @@
+                continue;
+            }
+            trackingLocalCheckpoint.put(initID, SequenceNumbersService.UNASSIGNED_SEQ_NO);
+            logger.trace("added [{}] the tracking map due to a CS update", initID);
```
nit: `added [{}] the` -> `added [{}] to the`

bleskes replied:
to added 👍

jasontedor commented on core/src/main/java/org/elasticsearch/index/seqno/LocalCheckpointService.java:
```diff
@@ -35,10 +36,8 @@
-    /** default value for {@link #SETTINGS_BIT_ARRAYS_SIZE} */
-    final static int DEFAULT_BIT_ARRAYS_SIZE = 1024;
+    public static Setting<Integer> SETTINGS_BIT_ARRAYS_SIZE = Setting.intSetting("index.seq_no.checkpoint.bit_arrays_size", 1024,
+        4, Setting.Property.IndexScope);
```
Why 4 and not 8? The bit set is backed by an array of longs?

bleskes replied:
I reduce it in testing to something small so I can more easily create multi chunk situations. It's not really based on efficiency (everything so low is not a good idea for production anyway). Also note that these are bits, not bytes.

jasontedor replied:
> Also note that these are bits, not bytes.

Yes, hence my comment that it's wasteful since 4 bits will require one long which could back 8 bits.

jasontedor commented on core/src/test/java/org/elasticsearch/action/support/replication/TransportReplicationActionTests.java:
```diff
@@ -519,9 +519,12 @@ public void testReplicaProxy() throws InterruptedException, ExecutionException {
+            final TransportReplicationAction.ReplicaResponse response =
+                new TransportReplicationAction.ReplicaResponse(randomAsciiOfLength(10), randomLong());
+            transport.handleResponse(captures[0].requestId,
+                response);
```
Nit: this can go on the previous line

bleskes replied:
moved.

jasontedor commented on core/src/test/java/org/elasticsearch/index/seqno/CheckpointsIT.java:
```diff
@@ -0,0 +1,80 @@
+import static org.hamcrest.Matchers.anyOf;
+import static org.hamcrest.Matchers.equalTo;
+
+@TestLogging("index.shard:TRACE,index.seqno:TRACE")
```
Do you think these are still needed, or were they just helpful during development?

bleskes replied:
Can remove. My logic was that in almost all cases you'd like to have this information when debugging. Let me know if you prefer it removed and I will.

jasontedor replied:
If your intuition is that these will be almost always needed, then obviously we should keep them.

jasontedor commented on core/src/test/java/org/elasticsearch/index/seqno/CheckpointsIT.java:
```diff
@@ -0,0 +1,80 @@
+            IndicesStatsResponse stats = client().admin().indices().prepareStats("test").clear().get();
+            for (ShardStats shardStats : stats.getShards()) {
+                if (shardStats.getSeqNoStats() == null) {
+                    assertFalse("didn't get seq no stats for a primary " + shardStats.getShardRouting(),
```
A simple change to the message can make this assertion fit on one line:

```diff
diff --git a/core/src/test/java/org/elasticsearch/index/seqno/CheckpointsIT.java b/core/src/test/java/org/elasticsearch/index/seqno/CheckpointsIT.java
index 5ccb541..ca172f8 100644
--- a/core/src/test/java/org/elasticsearch/index/seqno/CheckpointsIT.java
+++ b/core/src/test/java/org/elasticsearch/index/seqno/CheckpointsIT.java
@@ -54,8 +54,7 @@ public class CheckpointsIT extends ESIntegTestCase {
             IndicesStatsResponse stats = client().admin().indices().prepareStats("test").clear().get();
             for (ShardStats shardStats : stats.getShards()) {
                 if (shardStats.getSeqNoStats() == null) {
-                    assertFalse("didn't get seq no stats for a primary " + shardStats.getShardRouting(),
-                        shardStats.getShardRouting().primary());
+                    assertFalse("no seq_no stats for primary " + shardStats.getShardRouting(), shardStats.getShardRouting().primary());
                     continue;
                 }
                 logger.debug("seq_no stats for {}: {}", shardStats.getShardRouting(),
```

bleskes replied:
sure. thx.

bleskes replied:
sure. thx.

jasontedor commented on core/src/test/java/org/elasticsearch/index/seqno/GlobalCheckpointTests.java:
```diff
@@ -0,0 +1,137 @@
+    }
+
+    public void testEmptyShards() {
+        assertFalse("no started shards means we should update the checkpoint", checkpointService.updateCheckpointOnPrimary());
```
Is there a "not" missing from this message?

bleskes replied:
yeah, that's messy. re-wrote

jasontedor commented on core/src/test/java/org/elasticsearch/index/seqno/GlobalCheckpointTests.java:
```diff
@@ -0,0 +1,137 @@
+            } else if (randomBoolean()) {
+                insync.add(id);
+                maxLocalCheckpoint = Math.min(maxLocalCheckpoint, localCheckpoint);
+            } else {
```
Since we go through this loop so few times, do you think it'd be better if we had a uniform distribution over the three cases instead of 0.5/0.25/0.25?

bleskes replied:
yeah, I tried to give it a more realistic distribution of real life. Will change to diversify the distribution.

jasontedor commented on core/src/main/java/org/elasticsearch/index/IndexService.java:
```diff
@@ -718,6 +725,33 @@ private void maybeRefreshEngine() {
+                            try {
+                                shard.updateGlobalCheckpointOnPrimary();
+                            } catch (EngineClosedException | AlreadyClosedException ex) {
+                                // fine - continue;
```
Can you leave a more elaborate comment as to why this is fine?

bleskes replied:
added

jasontedor commented on core/src/main/java/org/elasticsearch/index/shard/ShadowIndexShard.java:
```diff
@@ -102,4 +103,22 @@ public boolean shouldFlush() {
+
+    @Override
+    public void updateGlobalCheckpointOnReplica(long checkpoint) {
+        // nocommit: think shadow replicas through
```
I assume the plan is to address this in a later PR (either way is fine, just want to know what the plan is)?

bleskes replied:
yeah. I want to take some time to think about the whole push in the light of seq_no, but I'm planning to let "normal" shard work to stabilize first.

jasontedor replied:
That makes sense.

jasontedor commented:
Since this has gone through a few rounds of review before, I largely have the same thoughts I had then. I gave it a thorough re-read and left some comments.

bleskes commented:
@jasontedor thx for the review. I pushed another commit addressing it.

jasontedor commented:
LGTM.

jasontedor commented:
LGTM.
